### PR TITLE
routing: add TlvTrafficShaper to bandwidth hints

### LINKED ISSUE #8617: routing: add custom sender-sider bandwidth hint selection as a top level config option
For certain custom channel types, we'll wan to extend the default [`bandwidthHints`](https://github.com/lightningnetwork/lnd/blob/master/routing/bandwidth.go#L10-L22) that the router uses to introspect into the local balance of a channel to make decisions w.r.t which channel should be selected as an outgoing edge. As an example, given a TLV blob contained in an `update_add_htlc` message, and a custom blob that may be stored along side an `OpenChannel/ChannleCommitment` struct, even though channels $`{X, Y, Z}`$ may have enough bandwidth to satisfy a payment of amount $`M`$, we may actually want to take channel $`A`$ instead, as $`F(htlc, A) == true`$. 

In this case, we want the creator of a new lnd process in an existing go program to be able to provide the function $`F`$ to us.

When we decide to make a payment, we'll make the following series of calls:
  * We'll call `getOutgoingBalance` to see if any of our local channels can even carry the payment: https://github.com/lightningnetwork/lnd/blob/master/routing/pathfind.go#L602
  * That in turn will call `getBalance` to see if the link can actually forward, and if so, return the active balance: https://github.com/lightningnetwork/lnd/blob/master/routing/bandwidth.go#L10-L22
      * This is the first instance that we'll want our cut out to apply. First, we'll need to extend `getBandwidth` to optionally be aware of the TLV blob of a given HTLC. Then, we'll want to just return zero (similar to if `!link.EligibleToForward()`), if our function $`F`$ returns false. 
   * Finally once we decide that some of our channels can carry the HTLC, we'll call `getEdgeLocal`: https://github.com/lightningnetwork/lnd/blob/b1175514f9854b965d4716ac015ea15387c63081/routing/unified_edges.go#L215. 
      * This is the second instance where our cut out will apply. Unconditionally, if $`F`$ returns true for a channel, we want to return the bandwidth. Otherwise, skip it if $`F`$ is present (could be `fn.Option`), but returns false. 

Here's an interface sketch that may be useful: 
```go
type TlvTrafficShaper interface {
    ShouldCarryPayment(amt lnwire.Millisatoshi, htlcTLV, channelBlob tlv.Blob) bool
}
```

The `htlcTLV` will just be what's present in the `update_add_htlc`, tho we'll also need a way to pass in a custom instance as path finding hasn't completed yet. `channelBlob` will be available based on the channels in the link.

### PR DESCRIPTION
**Pull request by GeorgeTsagk**
## Description

Closes #8617 

Based on #8660

This PR adds a `TlvTrafficShaper` function as part of the bandwidth hints. This function can be optionally defined to provide an external way of evaluating whether a channel is capable of carrying out a payment or not.

### CODE REVIEW

File: routing/bandwidth.go [hunk 1]
```diff
@@ -18,7 +20,20 @@ type bandwidthHints interface {
 	// will be used. If the channel is unavailable, a zero amount is
 	// returned.
 	availableChanBandwidth(channelID uint64,
-		amount lnwire.MilliSatoshi) (lnwire.MilliSatoshi, bool)
+		amount lnwire.MilliSatoshi,
+		blobs ...tlv.Blob) (lnwire.MilliSatoshi, bool)
```
Comment (Roasbeef) [inline routing/bandwidth.go:24]: Is this the HTLC blob? Or all the channel blobs? Could be a `fn.Option[tlv.Blob]` if it's the HTLC blobs.

File: routing/bandwidth.go [hunk 2]
```diff
@@ -89,30 +107,71 @@ func (b *bandwidthManager) getBandwidth(cid lnwire.ShortChannelID,
...
+		}
+
+		// Otherwise, we'll return the current best estimate for the
+		// available bandwidth for the link.
+		return link.Bandwidth()
+	}
+
+	// If the provided tlv blobs are not of the expected number, then we'll
+	// log a warning and return zero bandwidth.
+	if len(blobs) != 2 {
```
Comment (Roasbeef) [inline routing/bandwidth.go:126]: Can we modify the type signature to more explicitly encode some of the desired semantics?

File: routing/bandwidth.go [hunk 3]
```diff
@@ -89,30 +107,71 @@ func (b *bandwidthManager) getBandwidth(cid lnwire.ShortChannelID,
...
+	}
+
+	htlcTlv := blobs[0]
+	channelBlob := blobs[1]
+
+	res := false
+
+	// If the traffic shaper is present, we'll check if the payment should
+	// be carried by the channel.
+	b.trafficShaper.WhenSome(func(ts TlvTrafficShaper) {
```
Comment (Roasbeef) [inline routing/bandwidth.go:139]: Slightly diff way to write this: 
```go
res := fn.MapOptionZ(b.trafficShaper, func(ts TlvTrafficShaper) bool {
    return ts.ShouldCarryPayment(amount, htlcTLv, channelBlob)
})
```

`MapOptionZ` will return the default value of the return arg (false in this case) if the option param is none.

File: routing/bandwidth.go [hunk 4]
```diff
@@ -89,30 +107,71 @@ func (b *bandwidthManager) getBandwidth(cid lnwire.ShortChannelID,
...
+	})
+
+	if !res {
+		log.Warnf("ShortChannelID=%v: payment not allowed by traffic "+
+			"shaper", cid)
 		return 0
 	}
 
-	// Otherwise, we'll return the current best estimate for the available
-	// bandwidth for the link.
```
Comment (Roasbeef) [inline routing/bandwidth.go:101]: I think we might actually want to return sort of "infinite" bandwidth here instead.

File: routing/bandwidth.go [hunk 5]
```diff
@@ -89,30 +107,71 @@ func (b *bandwidthManager) getBandwidth(cid lnwire.ShortChannelID,
...
 
-	return b.getBandwidth(shortID, amount), true
+	return b.getBandwidth(shortID, amount, blobs...), true
+}
+
+// hasTrafficShaper returns true if the bandwidth manager has a traffic shaper.
+func (b *bandwidthManager) hasTrafficShaper() bool {
+	res := false
+
+	b.trafficShaper.WhenSome(func(_ TlvTrafficShaper) {
```
Comment (Roasbeef) [inline routing/bandwidth.go:172]: Can be written as: 
```go
return b.trafficShaper.IsSome()
```

File: routing/router.go [hunk 6]
```diff
@@ -405,6 +407,8 @@ type Config struct {
 	// IsAlias returns whether a passed ShortChannelID is an alias. This is
 	// only used for our local channels.
 	IsAlias func(scid lnwire.ShortChannelID) bool
+
+	TrafficShaper fn.Option[TlvTrafficShaper]
```
Comment (Roasbeef) [inline routing/router.go:411]: We also want to push this all the way up to `lnd.ImplConfig`.

File: routing/unified_edges.go [hunk 7]
```diff
@@ -164,10 +165,10 @@ type edgeUnifier struct {
 // specific amount to send. It differentiates between local and network
 // channels.
 func (u *edgeUnifier) getEdge(amt lnwire.MilliSatoshi,
-	bandwidthHints bandwidthHints) *unifiedEdge {
+	bandwidthHints bandwidthHints, blobs ...tlv.Blob) *unifiedEdge {
```
Comment (Roasbeef) [inline routing/unified_edges.go:168]: Where are these set?

File: server.go [hunk 8]
```diff
@@ -988,6 +989,7 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 		Clock:               clock.NewDefaultClock(),
 		StrictZombiePruning: strictPruning,
 		IsAlias:             aliasmgr.IsAlias,
+		TrafficShaper:       fn.None[routing.TlvTrafficShaper](),
```
Comment (Roasbeef) [inline server.go:992]: See the comment about `ImplConfig`.

**Review by Roasbeef — CHANGES_REQUESTED** [review]

**Comment by dstadulis** [comment]
Was unable to determine why GH isn't linking this PR to the issue

>Closes https://github.com/lightningnetwork/lnd/issues/8617

File: routing/bandwidth.go (hunk 4 above)
Comment (GeorgeTsagk) [reply routing/bandwidth.go:101]: hmm given that this method is also used in `getOutgoingBalance` should I provide nil blobs when coming from that path?

File: routing/bandwidth.go [hunk 9]
```diff
@@ -89,30 +130,86 @@ func (b *bandwidthManager) getBandwidth(cid lnwire.ShortChannelID,
...
+		)
+		if err != nil {
+			return fmt.Errorf("failed to get bandwidth from "+
+				"external traffic shaper: %w", err)
+		}
+
+		log.Debugf("ShortChannelID=%v: external traffic shaper "+
+			"reported available bandwidth: %v", cid, auxBandwidth)
+
+		// TODO(guggero): Still need to ask the link if it can add an
```
Comment (guggero) [inline routing/bandwidth.go:167]: This was answered by @Roasbeef here: https://github.com/lightningnetwork/lnd/commit/17a05186d325d648e1a642a254ff1c1a999015c0#r141783834.
So I think it's fine to remove the comment.

File: routing/bandwidth.go (hunk 4 above)
Comment (guggero) [reply routing/bandwidth.go:101]: This is solved by the current version by returning the bandwidth the external traffic shaper returns.

File: routing/payment_session_source.go [hunk 10]
```diff
@@ -42,6 +43,8 @@ type SessionSource struct {
 	// PathFindingConfig defines global parameters that control the
 	// trade-off in path finding between fees and probability.
 	PathFindingConfig PathFindingConfig
+
+	TrafficShaper fn.Option[TlvTrafficShaper]
```
Comment (guggero) [inline routing/payment_session_source.go:47]: nit: needs Godoc comment.
Can copy the one from the interface definition itself:
```
// TlvTrafficShaper is an interface that allows the sender to determine if a
// payment should be carried by a channel based on the TLV records that may be
// present in the `update_add_htlc` message or the channel commitment itself.
```

File: routing/pathfind.go [hunk 11]
```diff
@@ -1029,9 +1047,25 @@ func findPath(g *graphParams, r *RestrictParams, cfg *PathFindingConfig,
 				continue
 			}
 
+			firstHopTLVs :=
+				tlv.MapToRecords(r.FirstHopCustomRecords)
```
Comment (guggero) [inline routing/pathfind.go:1051]: nit: formatting (break after parenthesis, same below.

File: routing/pathfind.go [hunk 12]
```diff
@@ -1223,3 +1257,12 @@ func lastHopPayloadSize(r *RestrictParams, finalHtlcExpiry int32,
 	// The final hop does not have a short chanID set.
 	return finalHop.PayloadSize(0)
 }
+
+func safeAdd(x, y uint64) uint64 {
```
Comment (guggero) [inline routing/pathfind.go:1261]: Should we make this method take and return `lnwire.MilliSatoshi` instead? Then the call itself is a bit slimmer. Also, would nice to add a Godoc comment. And maybe "overflowSafeAdd" would be more fitting?

File: routing/pathfind.go [hunk 13]
```diff
@@ -1223,3 +1257,12 @@ func lastHopPayloadSize(r *RestrictParams, finalHtlcExpiry int32,
...
+	if y > math.MaxUint64-x {
+		// Overflow would occur, return maximum uint64 value
```
Comment (guggero) [inline routing/pathfind.go:1263]: nit: missing full stop in comment.

**Review by guggero — COMMENTED** [review]
Nice work, LGTM :tada: 
Just nits remaining

**Review by guggero — APPROVED** [review]
LGTM :tada:

**Pull request merged**